### PR TITLE
Local WDL storage and downloading

### LINKED ISSUE #137: Storing WDLs locally
Instead of referencing WDLs by their URI, we should store them locally somewhere.  This will be the behavior for the stripped down docker-compose that is made to run locally, which will mean we need to make sure wherever we put those WDLs in a volume.

### LINKED ISSUE #138: Downloading WDLs
Since we'll be storing WDLs somewhere, possibly locally depending on your configuration, we should add some functionality to the REST API to allow downloading the WDLs.

### PR DESCRIPTION
**Pull request by KevinCLydon**
Added a step in template creation and updating to download and store wdls locally. Added REST mappings for retrieving the WDLs for a template. Modified what is returned by calls to Template REST mappings to display the REST mapping for a WDL if it is stored locally, as opposed to the local file path. For WDLs that are currently still referenced in the TEMPLATE table with a gs, http, or https uri, that uri will still be returned.

This ended up being two issues (closes #137 and closes #138) because I realized while writing it that we would be exposing the local paths of the WDLs to the user if I just did the WDL storage first, and I didn't feel great about that.  Instead, I added mappings for downloading WDLs, and now, if you search for a template, the values for `test_wdl` and `eval_wdl` will be the respective REST URLs for downloading.

I'm undecided on a couple minor aspects of this:
- Right now, if you hit the download mapping for a template's test or eval wdl, and the WDL is stored in GCS (meaning it wasn't downloaded at template creation for some reason (e.g. the template was created using a previous CARROT version, the template was created when a CARROT was configured to store WDLs in GCS, etc.), CARROT will download the WDL from GCS and then send it to the user.  I'm not sure if this is a potential issue.  It could allow a user access to a WDL they wouldn't normally be able to access, because CARROT can access it.  Although, they could also access it through the Cromwell instance, so I'm not sure if this is a huge deal.
- I'm not married to the idea of hiding local file paths.  It seemed like the right choice, since I believe sharing local file paths is generally frowned upon from a security standpoint, but maybe not within the context CARROT is likely to be used?
- The mapping for downloading a test_wdl is `/api/v1/templates/{template_id}/test.wdl`.  Should we change it to `test_wdl`?  Maybe some other name would be better?
- Generating the WDL download URI to send back to the user feels slightly hacky as it is now.  Mostly because we get the host from the request the user sends, and also because the /api/v1 part of the URI is hardcoded, since that is added to REST mappings outside of the `routes` module.  I'm not sure of a better solution, but I'm happy to spend some more time looking for one, if we think this is a sufficiently large problem.

### CODE REVIEW

File: src/routes/template.rs [hunk 1]
```diff
@@ -140,15 +177,32 @@ async fn create(
         is_gs_uris_for_wdls_enabled()?;
     }
 
-    // Start by validating the WDLs
-    validate_wdl(&client, &new_template.test_wdl, "test", &new_template.name).await?;
-    validate_wdl(&client, &new_template.eval_wdl, "eval", &new_template.name).await?;
+    let conn = pool.get().expect("Failed to get DB connection from pool");
+
+    // Store and validate the WDLs
+    let test_wdl_location = store_wdl(&client, &conn, &new_template.test_wdl, WdlType::Test).await?;
```
Comment (jonn-smith) [inline src/routes/template.rs:183]: If the WDLs don't validate I feel like we shouldn't store them at all and should just return an error.

What do you think?

**Review by jonn-smith — CHANGES_REQUESTED** [review]
Looks good!

I had a question and a couple of comments.

With regard to your questions:

1. I don't think it's an issue that CARROT forwards a WDL from a bucket.  If CARROT can get at it, I think we can assume that it's not private.  If this ends up being a problem in the future we can revisit it.
2. I like hiding local file paths both for security reasons and because the path can potentially change, so it doesn't have any actual meaning.
3. I like `test_wdl` better than `test.wdl`.  It makes it seem more like an API end point than a file.
4. I think we can leave the link generation as is for now.  Can you make an issue to revisit it later?

File: src/models/wdl_hash.rs [hunk 2]
```diff
@@ -0,0 +1,197 @@
...
+
+/// A new wdl_hash record to be inserted into the DB
+///
+/// location and data_to_hash are required fields, and created_at is populated automatically by the
+/// DB
+/// Before insertion, `data_to_hash` is hashed using SHA-512 and the hash is what is actually
+/// inserted into the DB
+pub struct NewWdlHash {
+    pub location: String,
+    pub data_to_hash: String,
```
[the remark this answers is not in the document]
Comment (KevinCLydon) [reply src/models/wdl_hash.rs:31]: Yup, I'm on board with that

File: src/models/wdl_hash.rs [hunk 3]
```diff
@@ -0,0 +1,197 @@
...
+    pub created_at: NaiveDateTime,
+}
+
+/// A new wdl_hash record to be inserted into the DB
+///
+/// location and data_to_hash are required fields, and created_at is populated automatically by the
+/// DB
+/// Before insertion, `data_to_hash` is hashed using SHA-512 and the hash is what is actually
+/// inserted into the DB
+pub struct NewWdlHash {
```
[the remark this answers is not in the document]
Comment (KevinCLydon) [reply src/models/wdl_hash.rs:29]: `WdlDataToHash` maybe?

[the remark this answers is not in the document]
Comment (jonn-smith) [reply src/models/wdl_hash.rs:29]: That works!

File: src/routes/template.rs (hunk 1 above)
Comment (KevinCLydon) [reply src/routes/template.rs]: That makes sense to me.  I have a fix for that now

**Review by jonn-smith — APPROVED** [review]
👍

**Pull request merged**